[Payment-Request] Notify routes

Prometee opened the pull request:
| Q               | A
|-----------------|-----
| Branch?         | 2.0
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | none
| License         | MIT

This PR add two new routes :

```yaml
# Classic notify
sylius_payment_method_notify:
    path: /payment-methods/{code}

# Notify using an existing PaymentRequest
sylius_payment_request_notify:
    path: /payment-requests/{hash}
```

Prometee committed:
Remove remaining empty file

Prometee committed:
Add 2 webhooks routes

NoResponseMate commented on src/Sylius/Bundle/PaymentBundle/Resources/config/routing/integrations/sylius.yaml:
```diff
@@ -0,0 +1,9 @@
+sylius_shop_webhook_payment_method:
```
```suggestion
sylius_webhook_payment_method:
```
or
```suggestion
sylius_payment_webhook_payment_method:
```

NoResponseMate commented on src/Sylius/Bundle/PaymentBundle/Provider/CompositeWebhookPaymentProvider.php:
```diff
@@ -0,0 +1,40 @@
+
+namespace Sylius\Bundle\PaymentBundle\Provider;
+
+use LogicException;
```
```suggestion
```

NoResponseMate commented on src/Sylius/Bundle/PaymentBundle/Resources/config/services/provider.xml:
```diff
@@ -43,5 +43,9 @@
+
+        <service id="sylius.provider.webhook_payment.composite" class="Sylius\Bundle\PaymentBundle\Provider\CompositeWebhookPaymentProvider">
+            <argument type="tagged_iterator" tag="sylius.payment_request.webhook.payment_provider" />
+        </service>
```
```suggestion
        <service id="sylius.provider.webhook_payment" class="Sylius\Bundle\PaymentBundle\Provider\CompositeWebhookPaymentProvider">
            <argument type="tagged_iterator" tag="sylius.payment_request.webhook.payment_provider" />
        </service>
```
An alias for the interface would be nice as well

Prometee committed:
Remove shop prefix

Co-authored-by: Jan Góralski <jan.wojciech.goralski@gmail.com>

Prometee committed:
Remove use and add file header

Prometee committed:
Remove composite suffix

Co-authored-by: Jan Góralski <jan.wojciech.goralski@gmail.com>

diimpp commented on src/Sylius/Behat/Context/Ui/PaypalContext.php:
Empty file, please remove it fully.

Prometee replied:
It's the purpose of the related commit 😉

diimpp commented on src/Sylius/Bundle/PaymentBundle/Action/WebhookPaymentMethodAction.php:
We need to add action's action to the class name, like

> NotifyPaymentMethodAction

or 

> NotifyPaymentRequestByPaymentMethodAction
> NotifyUnsafePaymentRequestAction

or

> NotifyPamentMethodWebhookAction

Personally I would avoid calling it a webhook unless `symfony/webhook` is used.

diimpp commented on src/Sylius/Bundle/PaymentBundle/Action/WebhookPaymentMethodAction.php:
```diff
@@ -0,0 +1,65 @@
+        $paymentRequest->setAction(PaymentRequestInterface::ACTION_NOTIFY);
+        $paymentRequest->setPayload($request);
+
+        $this->paymentRequestRepository->add($paymentRequest);
```
I'm not up to date with payment request overall design, but as far as I see any http request to this controller/url will be automatically saved to DB?

How do I discard this request, if I don't consider it's correct, like signature verification at payum `NotifyNullAction`?

Prometee replied:
L53:
```php
$payment = $this->webhookPaymentProvider->getPayment($request, $paymentMethod);
```

This previous line is asking to the developer what is the related payment contains in the request, then only if a payment is returned the request will be saved to the database during a message processing (syn or async).
Then the responsability rely on the payment method developer.

Do you see another way of handling this ?

diimpp commented on src/Sylius/Bundle/PaymentBundle/Provider/CompositeWebhookPaymentProvider.php:
```diff
@@ -0,0 +1,48 @@
+        Assert::allIsInstanceOf($this->webhookPaymentProviders, WebhookPaymentProviderInterface::class);
+    }
+
+    public function getPayment(Request $request, PaymentMethodInterface $paymentMethod): PaymentInterface {
```
Please apply CS fixes.

```suggestion
    public function getPayment(Request $request, PaymentMethodInterface $paymentMethod): PaymentInterface
    {
```

diimpp commented on src/Sylius/Bundle/PaymentBundle/Provider/WebhookPaymentProviderInterface.php:
```diff
@@ -0,0 +1,15 @@
+use Sylius\Component\Payment\Model\PaymentMethodInterface;
+use Symfony\Component\HttpFoundation\Request;
+
+interface WebhookPaymentProviderInterface {
```
```suggestion
interface WebhookPaymentProviderInterface
{
```

diimpp commented on src/Sylius/Bundle/PaymentBundle/Resources/config/services/actions.xml:
```diff
@@ -0,0 +1,36 @@
+    <services>
+        <defaults public="true" />
+
+        <service id="sylius.action.webhook_payment_request" class="Sylius\Bundle\PaymentBundle\Action\WebhookPaymentRequestAction">
```
I think services still should be called controllers.

```suggestion
        <service id="sylius.controller.webhook_payment_request" class="Sylius\Bundle\PaymentBundle\Action\WebhookPaymentRequestAction">
```

```suggestion
        <service id="sylius.controller.payment.notify_payment_request" class="Sylius\Bundle\PaymentBundle\Controller\NotifyPaymentRequestAction">
```

diimpp commented on src/Sylius/Bundle/PaymentBundle/Resources/config/services/provider.xml:
```diff
@@ -43,5 +43,9 @@
             <tag name="sylius.payment_request.http_response.provider" gateway-factory="offline" />
         </service>
+
+        <service id="sylius.provider.webhook_payment" class="Sylius\Bundle\PaymentBundle\Provider\CompositeWebhookPaymentProvider">
```
All services in this file marked as public -- does this service needs to be public?

diimpp commented on src/Sylius/Bundle/PaymentBundle/Provider/CompositeWebhookPaymentProvider.php:
```diff
@@ -0,0 +1,48 @@
+        ));
+    }
+
+    public function supports(Request $request, PaymentMethodInterface $paymentMethod): bool {
```
```suggestion
    public function supports(Request $request, PaymentMethodInterface $paymentMethod): bool
    {
```

diimpp commented on src/Sylius/Bundle/PaymentBundle/Provider/CompositeWebhookPaymentProvider.php:
```diff
@@ -0,0 +1,48 @@
+            }
+        }
+
+        throw new \LogicException(sprintf(
```
This exception cannot be logic since logic exception should lead to direct fix by a developer. 
And this exception can only be triggered on incorrect `code`, so it's more like

```suggestion
        throw new \InvalidArgumentException(sprintf(
```

Prometee replied:
This is the purpose of this excepton, ask the developer to found a way to retrieve the right payment from the current request and paymentMethod.

diimpp commented on src/Sylius/Bundle/PaymentBundle/Provider/CompositeWebhookPaymentProvider.php:
```diff
@@ -0,0 +1,48 @@
+
+    public function getPayment(Request $request, PaymentMethodInterface $paymentMethod): PaymentInterface {
+        foreach ($this->webhookPaymentProviders as $provider) {
+            if ($provider->supports($request, $paymentMethod)) {
```
Just thinking out load, this approach allows to register multiple provides, that supports the same payment method and only first one will be triggered -- Maybe we shouldn't allow this and throw error earlier.

Prometee replied:
It's not only the payment method, it can be a mix of info contains in the request and the payment method that allow/disallow supports.
Example with Stripe:
Webhooks are attached to a list of events, I can have multiple providers one for each event and I need to know what PaymentMethod I'm using and the current request to decide if the provider support it or not.

diimpp commented on src/Sylius/Bundle/PaymentBundle/Action/WebhookPaymentRequestAction.php:
```diff
@@ -0,0 +1,54 @@
+
+        $this->paymentRequestAnnouncer->dispatchPaymentRequestCommand($paymentRequest);
+
+        return new Response('', 204);
```
Different payment gateways require different response, so this needs to be configurable somehow. (Unless it happens via announcer). For example my most recent webhook for custom payum gateway required `'OK', 200` to be returned.

Prometee replied:
I though about this during the development and let it for another future PR.
I also encountered this kind of specific response for Monetico and I think about doing the same thing that I've done here : https://github.com/Sylius/Sylius/blob/payment-request/src/Sylius/Bundle/CoreBundle/OrderPay/Provider/PaymentRequestPayResponseProvider.php#L65-L71

diimpp commented on src/Sylius/Bundle/PaymentBundle/Resources/config/routing/integrations/sylius.yaml:
```diff
@@ -0,0 +1,9 @@
+sylius_webhook_payment_method:
+    path: /payment-method/{code}
```
What if I've payment gateway with static single webhook address, but multiple payment methods?

For example I've configured payment methods at sylius as 
> acme_cards
> acme_crypto
> acme_bank_transfer

but at acme merchant panel, I can only configure single webhook address, which will receive notifications for all three payment methods? I've experienced at least 3 payment gateways that do this in recent years.

Maybe it can be solved with something like
> /payment-factories/{code}

Prometee replied:
Interesting, this can be an aditional feature for another PR, the payment provider behavior can be replicate to a payment method provider depending only on the current request.

diimpp commented on src/Sylius/Bundle/PaymentBundle/Resources/config/routing/integrations/sylius.yaml:
```diff
@@ -0,0 +1,9 @@
+sylius_webhook_payment_method:
+    path: /payment-method/{code}
```
Should it be like so?
```suggestion
    path: /payment-methods/{code}
```

diimpp commented on src/Sylius/Bundle/PaymentBundle/Resources/config/routing/integrations/sylius.yaml:
```diff
@@ -0,0 +1,9 @@
+        _controller: sylius.action.webhook_payment_request
+
+sylius_shop_webhook_payment_request:
+    path: /payment-request/{hash}
```
Should it be like so?
```suggestion
    path: /payment-requests/{hash}
```

diimpp commented on src/Sylius/Bundle/PaymentBundle/Resources/config/routing/integrations/sylius.yaml:
```diff
@@ -0,0 +1,9 @@
+    defaults:
+        _controller: sylius.action.webhook_payment_request
+
+sylius_shop_webhook_payment_request:
```
```suggestion
sylius_payment_webhook_payment_request:
```
```suggestion
sylius_payment_webhook_notify_payment_request:
```
```suggestion
sylius_payment_notify_payment_request:
```

diimpp commented on src/Sylius/Bundle/PaymentBundle/Resources/config/routing/integrations/sylius.yaml:
```diff
@@ -0,0 +1,9 @@
+sylius_webhook_payment_method:
```
```suggestion
sylius_payment_webhook_payment_method:
```
```suggestion
sylius_payment_webhook_payment_notify_payment_method:
```
```suggestion
sylius_payment_notify_payment_method:
```

Prometee replied:
I did not wanted to replicate what Payum was doing (naming it Notify) but when I re-think of it most of the gateway I worked with are naming it notify url then I'm going to replace webhook by notify in all places.

Prometee committed:
Global renaming to Notify

# Conflicts:
#	src/Sylius/Bundle/PaymentBundle/Resources/config/services/controllers.xml

Prometee committed:
Add Attribute for autoconfiguration

Prometee committed:
Add plural to the route

Prometee committed:
Add a new service to wrap the non serializable Symfony request

Prometee committed:
Use symfony request wrapper

Prometee committed:
Do not re-run the same notify when it was already completed

Prometee committed:
Fix routes

Prometee committed:
Update service definition

Prometee committed:
Rename property

Prometee committed:
Add Dummy services to handle notify on offline payment method

Prometee committed:
Fix wrong order number generation

Prometee committed:
New context to handle Payment request notify urls

Prometee committed:
CS fix

Prometee committed:
Fix ecs

Prometee committed:
Array definition

diimpp commented on src/Sylius/Behat/Context/Setup/PaymentRequestContext.php:
```diff
@@ -55,7 +55,7 @@ public function thePaymentRequestActionHasBeenExecutedForOrderWithThePaymentMeth
      */
     public function thePaymentRequestActionHasBeenExecutedForOrderWithThePaymentMethodAndState(
-        string $state = null,
+        ?string $state,
         string $action,
         OrderInterface $order,
         PaymentMethodInterface $paymentMethod,
```
Nullable property should be placed after required properties.

```suggestion
        string $action,
        OrderInterface $order,
        PaymentMethodInterface $paymentMethod,
        ?string $state = null,
```

diimpp replied:
Though why `$state` is nullable? If it's null, then `$this->getTransitionForState($state)` will be called and it will crash since this function requires state to be string.

Prometee replied:
Apparently the error I fixed was fixed by another PR, the argument I changed was like that :
```php
string $state = null,
```
And this argument being the first one...

I will remove my commit since my fix is no longer needed, thank you for pointing it out 😉

Prometee committed:
Remove null type

Prometee committed:
Simplify if statement

Prometee committed:
Avoid using specific final state, instead check if it is one

Prometee committed:
Rename wrapper to normalizer

Prometee committed:
Move final transition checker to a dedicated service and add it to the announcer

Prometee committed:
Introduce a new service to handle request payload processing

Prometee committed:
Ease the autowiring

Prometee committed:
Missing file header

Prometee committed:
Add missing NotifyPaymentRequest handler for Payum

Prometee committed:
Introduce a new service to handle Notify response

NoResponseMate commented on src/Sylius/Bundle/PaymentBundle/Processor/RequestPayloadProcessor.php:
```diff
@@ -0,0 +1,41 @@
+        if (is_array($payload)) {
+            $payload += $requestPayload;
+            $paymentRequest->setPayload($payload);
+        }
```
```suggestion
        if (is_array($payload)) {
            $paymentRequest->setPayload($payload += $requestPayload);

            return;
        }
```
also guess it should be `array_merge` instead of `+=` since the payload from the request is fresher?

NoResponseMate commented on src/Sylius/Behat/Page/Shop/PaymentRequest/PaymentRequestNotifyPageInterface.php:
```diff
@@ -0,0 +1,39 @@
+
+interface PaymentRequestNotifyPageInterface extends BasePageInterface
+{
+    /**
+     * Calls a URI.
+     *
+     * @param string      $method        The request method
+     * @param array       $urlParameters The url parameters
+     * @param array       $files         The files
+     * @param array       $server        The server parameters (HTTP headers are referenced with an HTTP_ prefix as PHP does)
+     * @param string|null $content       The raw body data
+     */
+    public function openWithClient(
+        string $method,
+        array $urlParameters = [],
+        array $files = [],
+        array $server = [],
+        ?string $content = null,
+    ): void;
```
```suggestion
    public function openWithClient(
        string $method,
        array $urlParameters = [],
    ): void;
```
From what I can see only these 2 are used atm

Prometee replied:
@NoResponseMate 
I agree but I think about plugins using this. I already struggle in the past to make this working using Sylius Behat services. For example I will need to set server headers to the request to make Stripe works with this Behat Page service.

NoResponseMate replied:
Fair enough 👋
The docblock can go away though, the parameters' names are pretty much self-explanatory

NoResponseMate commented on src/Sylius/Bundle/PaymentBundle/Processor/RequestPayloadProcessor.php:
```diff
@@ -0,0 +1,41 @@
+
+        if (null === $payload) {
+            $paymentRequest->setPayload($requestPayload);
+        }
```
Since the `payload` is mixed and can basically be anything, shouldn't we have a composite in here? Just so you can provide a custom processor per gateway

NoResponseMate commented on src/Sylius/Bundle/PaymentBundle/Processor/RequestPayloadProcessor.php:
```diff
@@ -0,0 +1,41 @@
+
+        if (null === $payload) {
+            $paymentRequest->setPayload($requestPayload);
+        }
```
Since the `payload` is mixed and can basically be anything, shouldn't we have a composite in here? Just so you can provide a custom processor per gateway

Prometee replied:
@NoResponseMate

Now using `array_merge` and I added a test.

Not all payment methods will need this but I agree with you, it needs a composite here, to avoid too many custom decoration.

NoResponseMate commented on src/Sylius/Bundle/PaymentBundle/Processor/RequestPayloadProcessor.php:
```diff
@@ -0,0 +1,41 @@
+use Sylius\Component\Payment\Model\PaymentRequestInterface;
+use Symfony\Component\HttpFoundation\Request;
+
+final class RequestPayloadProcessor implements RequestPayloadProcessorInterface
```
```suggestion
final class NotifyPayloadProcessor implements NotifyPayloadProcessorInterface
```
seems more apt

NoResponseMate commented on src/Sylius/Bundle/PaymentBundle/Checker/PaymentRequestFinalTransitionChecker.php:
```diff
@@ -0,0 +1,38 @@
+use Sylius\Component\Payment\Model\PaymentRequestInterface;
+use Sylius\Component\Payment\PaymentRequestTransitions;
+
+final class PaymentRequestFinalTransitionChecker implements PaymentRequestFinalTransitionCheckerInterface
```
```suggestion
final class CompletedPaymentRequestChecker implements CompletedPaymentRequestCheckerInterface
```
A bit more ambiguity and the method would be `isCompleted`

Prometee replied:
@NoResponseMate From the state machine perspective it's a final state, "completed" rely to the actual payment request state but another state can be set by developers here.

Prometee replied:
```suggestion
final class PaymentRequestFinalStateChecker implements PaymentRequestFinalStateCheckerInterface
```

What about replacing "Transition" to "State" ?

NoResponseMate replied:
Ah, I forgot that there is an actual `completed` state. I meant it in a way that there is no more processing that can be done on a payment request, the fact that this implementation relies on a state machine is just an implementation detail. Maybe something like `FinalizedPaymentRequestChecker`? This suggestion applies more to the interface though.

Prometee replied:
I like it ;)

Let's rename both of them then

```suggestion
final class FinalizedPaymentRequestChecker implements FinalizedPaymentRequestCheckerInterface
```

NoResponseMate commented on src/Sylius/Bundle/PaymentBundle/Provider/PaymentNotifyProviderInterface.php:
```diff
@@ -0,0 +1,25 @@
+use Sylius\Component\Payment\Model\PaymentMethodInterface;
+use Symfony\Component\HttpFoundation\Request;
+
+interface PaymentNotifyProviderInterface
```
```suggestion
interface NotifyPaymentProviderInterface
```
```suggestion
interface NotifiedPaymentProviderInterface
```
or smth alike, the current name sounds like I'd put a payment inside and get a notification as the result

NoResponseMate commented on src/Sylius/Bundle/PaymentBundle/Processor/RequestPayloadProcessor.php:
```diff
@@ -0,0 +1,41 @@
+final class RequestPayloadProcessor implements RequestPayloadProcessorInterface
+{
+    public function __construct(
+        private SymfonyRequestNormalizerInterface $requestWrapper,
```
```suggestion
        private SymfonyRequestNormalizerInterface $requestNormalizer,
```

Prometee committed:
Rename RequestPayload processor to NotifyPayload and use array_merge

Prometee committed:
Add test for NotifyPayloadProcessor

Prometee committed:
Remove docblock

Prometee committed:
Rename PaymentNotify to NotifyPayment

Prometee committed:
Fix ECS

Prometee committed:
Rename PaymentRequestFinalTransitionChecker to FinalizedPaymentRequestChecker

GSadee commented on features/shop/payment_request/calling_notify_for_a_payment_method.feature:
```diff
@@ -0,0 +1,26 @@
+        And the store has a payment method "Offline" with a code "offline"
+        And there is a customer "john@example.com" that placed order with "PHP T-Shirt" product to "United States" based billing address with "Free" shipping method and "Offline" payment method
+
+    @ui
```
If I think correctly, this is neither `@api` nor `@ui`, so we could introduce a new `@webhook` tag here, but of course in a separate PR 🖖🏻

Prometee replied:
Yes, that's what I was thinking too. I did not test without tag but having a proper one is better.

GSadee commented on features/shop/payment_request/calling_notify_for_a_payment_method.feature:
```diff
@@ -0,0 +1,26 @@
+        When I call the payment method notify page with the code "not_existing"
+        Then no payment request with "notify" action should exists
+        And the response status code should be 404
+
```
```suggestion
```

GSadee commented on features/shop/payment_request/calling_notify_for_a_payment_request.feature:
```diff
@@ -0,0 +1,29 @@
+        And the store ships everywhere for Free
+        And the store has a payment method "Offline" with a code "offline"
+        And there is a customer "john@example.com" that placed order with "PHP T-Shirt" product to "United States" based billing address with "Free" shipping method and "Offline" payment method
+
+
```
```suggestion

```

GSadee commented on features/shop/payment_request/calling_notify_for_a_payment_request.feature:
```diff
@@ -0,0 +1,29 @@
+        When I call the payment request notify page for this payment request
+        Then a payment request with "notify" action and state "completed" should exists
+        And the response status code should be 404
+
```
```suggestion
```

GSadee commented on src/Sylius/Behat/Context/Ui/Shop/PaymentRequestContext.php:
```diff
@@ -0,0 +1,119 @@
+    }
+
+    /**
+     * @When /^I call the payment request notify page for this payment request$/
```
```suggestion
     * @When I call the payment request notify page for this payment request
```

GSadee commented on src/Sylius/Behat/Context/Ui/Shop/PaymentRequestContext.php:
```diff
@@ -0,0 +1,119 @@
+        Assert::count($paymentRequests, 1);
+        $paymentRequest = $paymentRequests[0];
+        Assert::notNull($paymentRequest);
+        Assert::eq($paymentRequest->getState(), $state);
```
```suggestion
        Assert::same($paymentRequest->getState(), $state);
```

GSadee commented on src/Sylius/Behat/Context/Ui/Shop/PaymentRequestContext.php:
```diff
@@ -0,0 +1,119 @@
+    }
+
+    /**
+     * @Then /^a payment request with "([^"]*)" action and state "([^"]*)" should exists$/
```
```suggestion
     * @Then a payment request with :action action and state :state should exist
```

GSadee commented on features/shop/payment_request/calling_notify_for_a_payment_request.feature:
```diff
@@ -0,0 +1,29 @@
+    Scenario: I want to send HTTP request to the payment method notify and failed
+        Given there is a "completed" "notify" payment request for order "#000001" using the "Offline" payment method
+        When I call the payment request notify page for this payment request
+        Then a payment request with "notify" action and state "completed" should exists
```
```suggestion
        Then a payment request with "notify" action and state "completed" should exist
```

GSadee commented on src/Sylius/Behat/Context/Ui/Shop/PaymentRequestContext.php:
```diff
@@ -0,0 +1,119 @@
+    }
+
+    /**
+     * @Given /^the response content should be empty$/
```
```suggestion
     * @Then the response content should be empty
```

GSadee commented on src/Sylius/Behat/Context/Ui/Shop/PaymentRequestContext.php:
```diff
@@ -0,0 +1,119 @@
+    }
+
+    /**
+     * @Given /^the response status code should be (\d+)$/
```
```suggestion
     * @Then the response status code should be :statusCode
```

Prometee replied:
Oups ! I did not see this one, I wonder why it did not get error though...

GSadee commented on src/Sylius/Behat/Context/Ui/Shop/PaymentRequestContext.php:
```diff
@@ -0,0 +1,119 @@
+    {
+        $response = $this->paymentMethodNotifyPage->getClient()->getInternalResponse();
+
+        Assert::eq($response->getStatusCode(), $statusCode);
```
```suggestion
        Assert::same($response->getStatusCode(), $statusCode);
```

GSadee commented on src/Sylius/Behat/Resources/config/suites/ui.yml:
```diff
@@ -42,6 +42,7 @@ imports:
     - ui/payment/managing_payment_methods.yml
     - ui/payment/managing_payments.yml
     - ui/product/navigating_between_product_show_and_edit_pages.yaml
+    - ui/payment_request/payment_request_notify.yml
```
Should be added according to the alphabetical order
```suggestion
    - ui/payment_request/payment_request_notify.yaml
```

Prometee replied:
@GSadee All the other files are yml and I think a more large work as to be done here on the entire repository

GSadee commented on src/Sylius/Behat/Service/PaymentRequest/Provider/DummyNotifyPaymentProvider.php:
```diff
@@ -0,0 +1,47 @@
+use Sylius\Component\Payment\Model\PaymentMethodInterface;
+use Symfony\Component\HttpFoundation\Request;
+
+#[AsNotifyPaymentProvider()]
```
```suggestion
#[AsNotifyPaymentProvider]
```
